prow/commentpruner/commentpruner.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package commentpruner facilitates efficiently deleting issue comments as a reaction to webhook events.
+package commentpruner
+
+import (
+	"sync"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+)
+
+// bufSize is the size of the sliding window and subsequently the cache size.
+const bufSize = 100
+
+type githubClient interface {
+	ListIssueComments(org, repo string, number int) ([]github.IssueComment, error)
+	DeleteComment(org, repo string, id int) error
+}
+
+// EventClient is a struct that provides comment deletion for an event related to an issue.
+// A single client instance should be created for each event and shared by all consumers of the event.
+// The client fetches the comments only once and filters that list repeatedly to find comments to
+// delete. This avoids using lots of API tokens when fetching comments for each handler that wants
+// to delete comments. (An HTTP cache only partially helps with this because deletions modify the
+// list of comments so the next call requires GH to send the resource again.)
+type EventClient struct {
+	org    string
+	repo   string
+	number int
+
+	ghc githubClient
+	log *logrus.Entry
+
+	once     sync.Once
+	lock     sync.Mutex
+	comments []github.IssueComment
+}
+
+// NewEventClient creates an EventClient struct. This should be used once per webhook event.
+func NewEventClient(ghc githubClient, log *logrus.Entry, org, repo string, number int) *EventClient {
+	return &EventClient{
+		org:    org,
+		repo:   repo,
+		number: number,
+
+		ghc: ghc,
+		log: log,
+	}
+}
+
+// PruneComments fetchest issue comments if they have not yet been fetched for this webhook event
+// and then deletes any indicated by the func 'shouldPrune'.
+func (c *EventClient) PruneComments(shouldPrune func(github.IssueComment) bool) {
+	c.once.Do(func() {
+		c.lock.Lock()
+		defer c.lock.Unlock()
```

Comment:
the locking here is unnecessary-- once provides its own lock

Reply:
The lock is there likely because `once` does not protect a second thread from accessing the comment slice.

Reply:
I think @rmmh  is right. All threads block on `c.once.Do` until the function has returned. See [here](https://golang.org/pkg/sync/#Once.Do).